src/audio/kpb.c

```diff
@@ -0,0 +1,541 @@
+ */
+static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
+{
+
+	struct comp_dev *dev;
+	struct sof_ipc_comp_kpb *kpb;
+	struct sof_ipc_comp_kpb *ipc_kpb = (struct sof_ipc_comp_kpb *)comp;
+	struct comp_data *cd;
+
+	trace_kpb("kpb_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_kpb->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KPB, ipc_kpb->config);
+		return NULL;
+	}
+
+	/* Validate input parameters */
+	if (ipc_kpb->channels > MAX_SUPPORETED_CHANNELS) {
+		trace_kpb_error("kpb_new() error: "
+		"nr of channels exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->history_depth > MAX_BUFFER_SIZE) {
+		trace_kpb_error("kpb_new() error: "
+		"history depth exceeded the limit");
+	return NULL;
+	}
+
+	if (ipc_kpb->sampling_freq != MAX_SAMPLNG_FREQUENCY) {
+		trace_kpb_error("kpb_new() error: "
+		"requested sampling frequency not supported");
+	return NULL;
+	}
+
+	if (ipc_kpb->sampling_width != SAMPLING_WIDTH) {
+		trace_kpb_error("kpb_new() error: "
+		"requested sampling width not supported");
+	return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		COMP_SIZE(struct sof_ipc_comp_kpb));
+	if (!dev)
+		return NULL;
+
+	kpb = (struct sof_ipc_comp_kpb *)&dev->comp;
+	memcpy(kpb, ipc_kpb, sizeof(struct sof_ipc_comp_kpb));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	cd->history_depth = ipc_kpb->history_depth;
+	comp_set_drvdata(dev, cd);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+}
+
+/**
+ * kpb_free() - reclaim memory for a key phrase buffer.
+ * @arg1: component device pointer
+ */
+static void kpb_free(struct comp_dev *dev)
+{
+	struct comp_data *kd = comp_get_drvdata(dev);
+
+	trace_kpb("kpb_free()");
+
+	rfree(kd);
+	rfree(dev);
+}
+
+static int kpb_trigger(struct comp_dev *dev, int cmd)
+{
+	trace_kpb("kpb_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/**
+ * kpb_prepare() - prepare key phrase buffer.
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_prepare(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+	int i;
+
+	trace_kpb("kpb_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	cd->no_of_clients = 0;
+
+	/* allocate history_buffer/s */
+#if KPB_NO_OF_HISTORY_BUFFERS == 2
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE - LPSRAM_SIZE);
+	cd->his_buf_lp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_LP,
+					  LPSRAM_SIZE);
+
+	if (cd->his_buf_hp.sta_addr || cd->his_buf_lp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+		return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
+	(MAX_BUFFER_SIZE - LPSRAM_SIZE);
+
+	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
+	LPSRAM_SIZE;
+
+#elif KPB_NO_OF_HISTORY_BUFFERS == 1
```

Comment:
Is there a possibility to do it more flexible? E.g. loop + array with proper values.

Reply:
It is possible however, that was prepared for kconfig to define KPB_NO_OF_HISTORY_BUFFERS as needed. So this macro:

`#define KPB_NO_OF_HISTORY_BUFFERS 2`

will be removed from kpb.h and handled by kconfig. 

I guess your proposal is to loop thru configured buffers and read/write data. It is also possible but the intention here was to not compile additional code for two buffer while we only have configured one.